### PR TITLE
[Merged by Bors] - feat(Combinatorics/Additive): characterise sets with no doubling

### PR DESCRIPTION
**Pull request by YaelDillies**
They are exactly the empty set and cosets of a subgroup.

From LeanCamCombi

Co-authored-by: Patrick Luo

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/Additive/NoDoubling.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/NoDoubling.lean:24]: Can we relate this hypothesis to one involving the doubling constant? I don't mind if it's in this file or in the doubling constant one (or even in a different PR), but we should have the statement d(A) = 1 iff #(A * A) <= #A somewhere

File: Mathlib/Combinatorics/Additive/NoDoubling.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/NoDoubling.lean:28]:
```suggestion
  have A_smul {a} (ha : a ∈ A) : A <• a = A * A :=
```
or something, to match the way it reads?

File: Mathlib/Combinatorics/Additive/NoDoubling.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/NoDoubling.lean:22]: Let's strengthen the docstring to match the statement more closely: "there exists a subgroup H of G such that every translate of H by an element of A is A itself). Even better would be to give the actual statement which this docstring is saying (as a corollary of this statement): if A is nonempty then there exists `a` with `aH = A`.

**Review by b-mehta — APPROVED** [review]

**Comment by b-mehta** [comment]
Also do we want to combine this file with the very small doubling stuff? It doesn't seem like there's a great deal to say about either notion (indeed your proof and my proof together show that there's nothing to say!), and they fit together nicely. So that'd suggest calling this file VerySmallDoubling or something, in anticipation, and it's not really inaccurate as a name anyway.

File: Mathlib/Combinatorics/Additive/NoDoubling.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/NoDoubling.lean:24]: I have this lemma proved locally in LeanCamCombi, but I was wondering whether it should live here or in the doubling constant file. We could just leave it to another PR for now, maybe

**Comment by YaelDillies** [comment]
Yes, very happy to merge both files

File: Mathlib/Combinatorics/Additive/NoDoubling.lean
Comment (b-mehta) [reply Mathlib/Combinatorics/Additive/NoDoubling.lean]: Yeah that's fine, let's just add a TODO for it. I don't have a strong preference where it should live, although this file should reference it if it doesn't contain it.

File: Mathlib/Combinatorics/Additive/VerySmallDoubling.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:12]: Only because you use the word "characterises" here, it would be nice to have the converse

Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:12]: The converse is already proved by material from the pointwise API: `∅ ^ 2 = ∅`, `(a • H) ^ 2 = a ^ 2 • H ^ 2 = a ^ 2 • H` (at least once you merge #19259 😇). I don't think it's really worth writing the converse explicitly?

File: Mathlib/Combinatorics/Additive/VerySmallDoubling.lean
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:17]:
```suggestion
* Do we need a version stated using the doubling constant (`Finset.mulConst`)?
* Add characterisation for sets with doubling < 3/2
```
just so the name makes sense in the meantime

File: Mathlib/Combinatorics/Additive/VerySmallDoubling.lean
Comment (b-mehta) [reply Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:12]: Agreed, but let's give a reference to those in that case. I still think it'd be kinda nice to have something of the form `#(A * A) ≤ #A` iff ... but I'm not too bothered

**Review by b-mehta — APPROVED** [review]
bors d+

File: Mathlib/Combinatorics/Additive/VerySmallDoubling.lean
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/VerySmallDoubling.lean:17]: Yep, makes sense

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**
